lib/src/socket.dart

```diff
@@ -392,211 +335,217 @@ class PhoenixSocket {
-        stacktrace,
       );
 
+      return false;
+    } catch (error, stackTrace) {
+      _logger.warning('Heartbeat message failed', error, stackTrace);
+      _reconnect(4001, reason: 'Heartbeat timeout');
       return false;
     }
   }
 
-  void _triggerChannelExceptions(PhoenixException exception) {
-    _logger.fine(
-      () => 'Trigger channel exceptions on ${channels.length} channels',
-    );
-    for (final channel in channels.values) {
-      _logger.finer(
-        () => 'Trigger channel exceptions on ${channel.topic}',
+  Timer _scheduleHeartbeat([String? previousHeartbeat]) {
+    return Timer(_options.heartbeat, () {
+      if (previousHeartbeat != null) {
+        final completer = _pendingMessages.remove(previousHeartbeat);
+        if (completer != null && !completer.isCompleted) {
+          completer.completeError(
+            TimeoutException(
+              'Heartbeat $previousHeartbeat not completed before sending new one',
+            ),
+            StackTrace.current,
+          );
+          return;
+        }
+      }
+
+      _sendHeartbeat();
+    });
+  }
+
+  void _cancelHeartbeat() {
+    if (_heartbeatTimeout != null) {
+      _heartbeatTimeout?.cancel();
+      _heartbeatTimeout = null;
+    }
+    if (_latestHeartbeatRef != null) {
+      _pendingMessages.remove(_latestHeartbeatRef)?.complete(
+            Message.heartbeat(_latestHeartbeatRef!), // doesn't matter
+          );
+      _latestHeartbeatRef = null;
+    }
+  }
+
+  bool _shouldPipeMessage(String message) {
+    final currentSocketState = _socketStateStream.valueOrNull;
+    if (currentSocketState is WebSocketReady) {
+      return true;
+    } else {
+      _logger.warning(
+        'Message from socket dropped because PhoenixSocket is not ready (is $currentSocketState)'
+        '\n$message',
       );
-      channel.triggerError(exception);
+
+      return false;
     }
   }
 
-  Message _heartbeatMessage() => Message.heartbeat(_nextHeartbeatRef = nextRef);
+  /// EVENT HANDLERS
+
+  // Exists for testing only
+  void onSocketDataCallback(String message) {
+    if (_shouldPipeMessage(message)) {
+      if (!_receiveStreamController.isClosed) {
+        _receiveStreamController.add(message);
+      }
+    }
+  }
 
   void _onMessage(Message message) {
-    _nextHeartbeatRef = null;
     if (message.ref != null) {
-      final completer = _pendingMessages[message.ref!];
+      _logger.finer('Received message with ref ${message.ref}');
+      final completer = _pendingMessages.remove(message.ref!);
       if (completer != null) {
-        _pendingMessages.remove(message.ref);
+        if (_logger.isLoggable(Level.FINEST)) {
+          _logger.finest('Completed ref ${message.ref} with $message');
+        }
         completer.complete(message);
       }
+
+      if (message.ref != _latestHeartbeatRef) {
+        // The connection is alive, prevent heartbeat timeout from closing
+        // connection.
+        _latestHeartbeatRef = null;
```

Comment:
Should we only clear this when the `message.ref != _latestHeartbeatRef` or should we just always clear the `_latestHeartbeatRef` when we receive any message as we do with the current implementation?

Reply:
Updated.